### PR TITLE
Fix `boundary_check` bug of `tl.make_block_ptr` for column-major for SGLANG

### LINKED ISSUE #6513: [SGLANG] `boundary_check` bug of `tl.make_block_ptr` for column-major
### Describe the bug

`boundary_check` in `tl.make_block_ptr` is silently ignored for column-major (order=(0,1)) 2D block pointers on Intel XPU Triton — OOB lanes return garbage (NaN) instead of 0. Row-major (order=(1,0)) works correctly. The bug affects both constexpr and runtime shape values.
This bug affects support of qwen3.5 model series in SGLang with triton kernels for linear attention.
Reproducer:
```python
import torch
import triton
import triton.language as tl


# ---------------------------------------------------------------------------
# Four kernels: {col,row}-major x {constexpr,runtime} T
# The column-major kernels expose the bug; row-major is the working control.
# ---------------------------------------------------------------------------

@triton.jit
def colmajor_constexpr(src, dst,
                       K: tl.constexpr, T: tl.constexpr, BT: tl.constexpr):
    """Column-major, constexpr T — FAILS on Intel XPU (OOB = garbage)."""
    blk = tl.make_block_ptr(
        src,
        shape=(K, T),
        strides=(1, K),       # K is the fast (stride-1) dimension
        offsets=(0, 0),
        block_shape=(K, BT),
        order=(0, 1),         # dimension 0 is contiguous → column-major
    )
    x = tl.load(blk, boundary_check=(1,))   # zero out cols >= T  (BROKEN on XPU)
    tl.store(dst + tl.arange(0, K)[:, None] * BT + tl.arange(0, BT)[None, :], x)


@triton.jit
def colmajor_runtime(src, dst,
                     K: tl.constexpr, T: tl.int32, BT: tl.constexpr):
    """Column-major, runtime T — FAILS on Intel XPU (OOB = garbage)."""
    blk = tl.make_block_ptr(
        src,
        shape=(K, T),         # T is a runtime value
        strides=(1, K),
        offsets=(0, 0),
        block_shape=(K, BT),
        order=(0, 1),
    )
    x = tl.load(blk, boundary_check=(1,))
    tl.store(dst + tl.arange(0, K)[:, None] * BT + tl.arange(0, BT)[None, :], x)


@triton.jit
def rowmajor_constexpr(src, dst,
                       K: tl.constexpr, T: tl.constexpr, BT: tl.constexpr):
    """Row-major, constexpr T — PASSES on Intel XPU (correct behaviour)."""
    blk = tl.make_block_ptr(
        src,
        shape=(K, T),
        strides=(BT, 1),      # T is the fast (stride-1) dimension
        offsets=(0, 0),
        block_shape=(K, BT),
        order=(1, 0),         # dimension 1 is contiguous → row-major
    )
    x = tl.load(blk, boundary_check=(1,))
    tl.store(dst + tl.arange(0, K)[:, None] * BT + tl.arange(0, BT)[None, :], x)


@triton.jit
def rowmajor_runtime(src, dst,
                     K: tl.constexpr, T: tl.int32, BT: tl.constexpr):
    """Row-major, runtime T — PASSES on Intel XPU (correct behaviour)."""
    blk = tl.make_block_ptr(
        src,
        shape=(K, T),
        strides=(BT, 1),
        offsets=(0, 0),
        block_shape=(K, BT),
        order=(1, 0),
    )
    x = tl.load(blk, boundary_check=(1,))
    tl.store(dst + tl.arange(0, K)[:, None] * BT + tl.arange(0, BT)[None, :], x)


# ---------------------------------------------------------------------------
# Test harness
# ---------------------------------------------------------------------------

def make_buffers(K, T, BT, device):
    """
    Column-major buffer: element (k, t) at offset k*1 + t*K.
    Valid range [0, K*T), rest filled with NaN (detectable garbage).
    """
    col_buf = torch.full((K * BT,), float("nan"), dtype=torch.float32, device=device)
    for t in range(T):
        col_buf[t * K : t * K + K] = 1.0

    """
    Row-major buffer: element (k, t) at offset k*BT + t.
    Each row k: first T entries = 1.0, rest = NaN.
    """
    row_buf = torch.full((K * BT,), float("nan"), dtype=torch.float32, device=device)
    for k in range(K):
        row_buf[k * BT : k * BT + T] = 1.0

    expected = torch.zeros(K, BT, device=device)
    expected[:, :T] = 1.0
    return col_buf, row_buf, expected


def run(device="xpu", K=4, BT=64, T=4):
    assert T < BT
    col_buf, row_buf, expected = make_buffers(K, T, BT, device)

    cases = [
        ("col-major constexpr T  [EXPECT FAIL on XPU]", colmajor_constexpr, col_buf),
        ("col-major runtime  T  [EXPECT FAIL on XPU]", colmajor_runtime,   col_buf),
        ("row-major constexpr T  [EXPECT PASS]",        rowmajor_constexpr, row_buf),
        ("row-major runtime  T  [EXPECT PASS]",         rowmajor_runtime,   row_buf),
    ]

    print(f"\nDevice={device}, K={K}, BT={BT}, T={T}  (OOB = cols {T}..{BT-1}, must be 0)\n")
    for name, fn, buf in cases:
        dst = torch.empty(K, BT, device=device)
        fn[(1,)](buf, dst, K=K, T=T, BT=BT)
        ok = torch.allclose(dst, expected, equal_nan=False)
        oob = dst[0, T:T+4].tolist()
        print(f"  {name}")
        print(f"    OOB sample = {oob}  PASS={ok}")

    col_bug = not torch.allclose(
        (lambda d: (colmajor_constexpr[(1,)](col_buf, d, K=K, T=T, BT=BT), d)[1])(
            torch.empty(K, BT, device=device)
        ),
        expected, equal_nan=False
    )
    row_ok = torch.allclose(
        (lambda d: (rowmajor_constexpr[(1,)](row_buf, d, K=K, T=T, BT=BT), d)[1])(
            torch.empty(K, BT, device=device)
        ),
        expected, equal_nan=False
    )
    if col_bug and row_ok:
        print("\n  *** BUG CONFIRMED: boundary_check broken for column-major 2D blocks ***")
    else:
        print("\n  (No bug observed — may be fixed in this triton version)")


if __name__ == "__main__":
    if torch.xpu.is_available():
        run("xpu")
    elif torch.cuda.is_available():
        run("cuda")
    else:
        print("No GPU available")
```
Assisted by copilot.

### Environment details

Triton version: triton-xpu 3.6.0
GPU: BMG B60

### PR DESCRIPTION
**Pull request by anmyachev**
E2E accuracy: https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/24985280889 (pass)

Flex attn: https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/24985314090 (pass)
Flex dec: https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/24985328173 (pass)

LTS run:
* https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/24990810226/job/73175656077 (wrong parameters)
* https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/24996592686/job/73195081715 (pass)

Benchmarks: https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/24997105056 (pass)

### CODE REVIEW

**Comment by anmyachev** [comment]
@chuanqi129 @etaf I don't see that you've opened a PR to update the Triton XPU version in PyTorch. Does this mean we have time to merge this change?

**Review by etiotto — DISMISSED** [review]

**Review by chengjunlu — DISMISSED** [review]

**Comment by chengjunlu** [comment]
The `order` of block pointer type (unlike the `order` of block layout. ) is meaningless and not used internally as we confirmed with OpenAI long ago. 
I think the fix is safe.

File: third_party/intel/lib/Dialect/Triton/Transforms/RemoveBoundaryChecks.cpp [hunk 1]
```diff
@@ -45,8 +45,16 @@ class BoundaryChecksRemover {
 
       SmallVector<int> newBoundaryCheck;
       for (int boundIdx : loadOp.getBoundaryCheck()) {
-        ArrayRef<int> order = makeTensorPtrOp.getOrder();
-        int idx = order.size() - order[boundIdx] - 1;
+        // `boundaryCheck` indices are in the same coordinate space as
+        // make_tensor_ptr shape/offset operands.
+        int idx = boundIdx;
+        if (idx < 0 || idx >= makeTensorPtrOp.getOffsets().size()) {
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Dialect/Triton/Transforms/RemoveBoundaryChecks.cpp]: wonder if line 51-57 is needed, is line 50 alone sufficient?

Comment (anmyachev) [reply third_party/intel/lib/Dialect/Triton/Transforms/RemoveBoundaryChecks.cpp]: looks like that

**Comment by anmyachev** [comment]
Please review it again.

**Review by chengjunlu — APPROVED** [review]

**Review by whitneywhtsang — APPROVED** [review]

**Comment by chuanqi129** [comment]
> @chuanqi129 @etaf I don't see that you've opened a PR to update the Triton XPU version in PyTorch. Does this mean we have time to merge this change?

Hi @anmyachev, triton for release branch don't need explicit upgrade commit pin PR, because it will take the latest commit of release branch when new pytorch RC tagged. According to the plan, today (this week) is the final RC tag day, current latest RC is RC3 (4 days ago). I think we have chance to land this PR, but we don't have change to fully test it, please make sure the risk of this PR is very very low.

**Comment by anmyachev** [comment]
> > @chuanqi129 @etaf I don't see that you've opened a PR to update the Triton XPU version in PyTorch. Does this mean we have time to merge this change?
> 
> Hi @anmyachev, triton for release branch don't need explicit upgrade commit pin PR, because it will take the latest commit of release branch when new pytorch RC tagged. According to the plan, today (this week) is the final RC tag day, current latest RC is RC3 (4 days ago). I think we have chance to land this PR, but we don't have change to fully test it, please make sure the risk of this PR is very very low.

got it, thanks @chuanqi129!
